File: practice_challenges/python/highest_palindrome.py
from test_case_data import test_case_palindrome, palindrome_result_correct

# from test_case_result import test_palindrome_result


def highestValuePalindrome(str_digits, str_length, num_changes) -> str:
    """Create highest value palindrome"""

    list_str = list(str_digits)
    mid_index = int(str_length / 2)
    start = 0
    end = str_length - 1
    stop_one = mid_index - 1
    stop_two = 0
    is_odd_str = str_length % 2 == 1
    is_odd_changes = num_changes % 2 == 1

    if is_odd_str:
        stop_two = mid_index + 1
        count = list_str.count("9")
        is_even_count = count % 2 == 0
        if is_odd_changes:
            if is_even_count:
                list_str[mid_index] = "9"
                num_changes -= 1
        else:
            if not is_even_count:
                list_str[mid_index] = "9"
                num_changes -= 1
    else:
        stop_two = mid_index

    while num_changes > 0:
        # print(num_changes)
        # print(start, end)
        digit_one = int(list_str[start])
        digit_two = int(list_str[end])

        if num_changes > 1:
            if digit_one != 9:
                list_str[start] = "9"
                num_changes -= 1
            if digit_two != 9:
                list_str[end] = "9"
                num_changes -= 1
        else:
            if digit_one < digit_two:
                list_str[start] = str(digit_two)
                num_changes -= 1
            elif digit_one > digit_two:
                list_str[end] = str(digit_one)
                num_changes -= 1
        # print(list_str)
        if start == stop_one or end == stop_two:
            break

        start += 1
        end -= 1

    # reversed_list = list(reversed(list_str))
    # is_equal = reversed_list == list_str
    # print(is_equal)
    # palindrome = "".join(reversed_list)
    # test_count = 0
    # for index in range(len(palindrome)):
    #     char_result_correct = palindrome_result_correct[index]
    #     char_result_test = palindrome[index]
    #     if char_result_correct != char_result_test:
    #         # print(mid_index)
    #         # print(index)
    #         # print(char_result_correct, char_result_test)
    #         break
    #         test_count += 1
    # # begin = 26078
    # begin = mid_index
    # print(len(test_case_palindrome))
    # print(len(palindrome_result_correct))
    # print(palindrome_result_correct[begin : begin + 100])
    # print(test_case_palindrome[begin : begin + 100])
    # print("-----------------------------------")
    # print(palindrome_result_correct[begin - 100 : begin + 1])
    # print(test_case_palindrome[begin - 100 : begin + 1])
    # return "testing"

    reversed_list = list(reversed(list_str))
    # print("reversed:", reversed_list)
    if reversed_list == list_str:
        palindrome = "".join(list_str)
        # print(palindrome)
        return palindrome

    else:
        return "-1"


# print(highestValuePalindrome("1231", 4, 3))
# print(highestValuePalindrome("12321", 5, 1))
# print(highestValuePalindrome("3943", 4, 1))
# print(highestValuePalindrome("092282", 6, 3))
# print(highestValuePalindrome("0011", 4, 1))
# print(highestValuePalindrome("932239", 6, 2))
# print(highestValuePalindrome(test_case_palindrome, 77543, 58343))

File: practice_challenges/typescript/highest_palindrome.ts
import { test_case_palindrome, test_case_palindrome_result } from './test_case_data'


const highestValuePalindrome = (str: string, strLen: number, numChanges: number): string => {
    const strArr = str.split('')
    const isOddStr = strLen % 2 === 1
    const mid = Math.floor(strLen / 2)
    let start = mid - 1
    let end = mid
    if (isOddStr) {
        end = mid + 1
    }

    let startIndex = 0
    let endIndex = strLen - 1
    while (true) {
        const isOddChanges = numChanges % 2 === 1
        const left = str[startIndex]
        const right = str[endIndex]
        const isNine = left !== '9' || right !== '9'
        const isEqual = left === right

        if (isOddStr && isEqual && numChanges === 1) {
            strArr[mid] = '9'
            numChanges--
            break
        }

        if (isNine) {
            if (left === '9') {
                strArr[endIndex] = '9'
                numChanges--
            } else if (right === '9') {
                strArr[startIndex] = '9'
                numChanges--
            } else if (isEqual && numChanges >= 2) {
                strArr[startIndex] = '9'
                strArr[endIndex] = '9'
                numChanges -= 2
            } else if (left > right) {
                if (numChanges >= 2) {
                    strArr[startIndex] = '9'
                    strArr[endIndex] = '9'
                    numChanges -= 2
                } else {
                    strArr[endIndex] = left
                    numChanges--
                }
            } else if (right > left) {
                if (numChanges >= 2) {
                    strArr[startIndex] = '9'
                    strArr[endIndex] = '9'
                    numChanges -= 2
                } else {
                    strArr[startIndex] = right
                    numChanges--
                }
            }

        }

        if (numChanges === 0 || startIndex === start || endIndex === end) {
            if (numChanges > 0) {
                if (isOddChanges && isOddStr) {
                    strArr[mid] = '9'
                    numChanges--
                }

                if (!isOddChanges) {
                    while (numChanges > 0) {
                        strArr[start] = '9'
                        strArr[end] = '9'
                        numChanges--
                        start--
                        end++
                    }
                }
            }
            break
        }

        startIndex++
        endIndex--
    }

    const reversedArr = [...strArr].reverse()
    const isPalindrome = strArr.every((char, index) => char === reversedArr[index])
    if (isPalindrome) {
        return strArr.join('')
    }

    return '-1'
}



// const highestValuePalindrome = (str: string, strLen: number, numChanges: number): string => {
//     const strArr = str.split('')
//     const isOddStr = strLen % 2 === 1
//     const mid = Math.floor(strLen / 2)
//     let start = mid - 1
//     let end = mid
//     if (isOddStr) end = mid + 1

//     let startIndex = start
//     let endIndex = end
//     while (true) {
//         if (numChanges === 0 || startIndex < 0 || endIndex > strLen - 1) break

//         const left = str[startIndex]
//         const right = str[endIndex]
//         const isEqual = left === right

//         if (numChanges === 1) {
//             if (isEqual) {
//                 // console.log('HERE1')
//                 if (isOddStr) {
//                     strArr[mid] = '9'
//                     numChanges--
//                 }
//             } else {
//                 if (left > right) {
//                     // console.log('HERE2')
//                     strArr[endIndex] = left
//                     numChanges--
//                 } else if (right > left) {
//                     // console.log('HERE3')
//                     strArr[startIndex] = right
//                     numChanges--
//                 }
//             }
//         } else if (numChanges === 2) {
//             if (left !== '9') {
//                 // console.log('HERE10')
//                 strArr[startIndex] = '9'
//                 numChanges--
//             }
//             if (right !== '9') {
//                 // console.log('HERE11')
//                 strArr[endIndex] = '9'
//                 numChanges--
//             }
//         } else if (numChanges > 2) {
//             // console.log('HERE4')
//             // if (!isEqual) {
//             if (left > right) {
//                 // console.log('HERE5')
//                 strArr[endIndex] = left
//                 numChanges--
//             } else if (right > left) {
//                 // console.log('HERE6')
//                 strArr[startIndex] = right
//                 numChanges--
//             }
//             // }
//         }


//         startIndex--
//         endIndex++
//     }

//     startIndex = 0
//     endIndex = strLen - 1
//     while (true) {
//         if (numChanges === 0 || startIndex > start || endIndex < end) break

//         const left = str[startIndex]
//         const right = str[endIndex]
//         const isEqual = left === right

//         if (isEqual) {
//             if (numChanges < 2) {
//                 // console.log('HERE7')
//                 if (isOddStr) {
//                     strArr[mid] = '9'
//                     numChanges--
//                 }

//             } else {
//                 if (left !== '9') {
//                     // console.log('HERE10')
//                     strArr[startIndex] = '9'
//                     numChanges--
//                 }
//                 if (right !== '9') {
//                     // console.log('HERE11')
//                     strArr[endIndex] = '9'
//                     numChanges--
//                 }
//             }

//         } else {
//             if (numChanges < 2) {
//                 if (left > right) {
//                     // console.log('HERE8')
//                     strArr[endIndex] = left
//                     numChanges--
//                 } else if (right > left) {
//                     // console.log('HERE9')
//                     strArr[startIndex] = right
//                     numChanges--
//                 }
//             } else {
//                 if (left !== '9') {
//                     // console.log('HERE10')
//                     strArr[startIndex] = '9'
//                     numChanges--
//                 }
//                 if (right !== '9') {
//                     // console.log('HERE11')
//                     strArr[endIndex] = '9'
//                     numChanges--
//                 }
//             }
//         }
//         startIndex++
//         endIndex--
//     }

//     const reversedArr = [...strArr].reverse()
//     const isPalindrome = strArr.every((char, index) => char === reversedArr[index])

//     if (isPalindrome) {
//         const result = strArr.join('')
//         return result
//     }

//     return '-1'
// }


console.log(highestValuePalindrome('1231', 4, 3))
console.log('===================')
console.log(highestValuePalindrome('12321', 5, 1))
console.log('===================')
console.log(highestValuePalindrome('3943', 4, 1))
console.log('===================')
console.log(highestValuePalindrome('092282', 6, 3))
console.log('===================')
console.log(highestValuePalindrome('0011', 4, 1))
console.log('===================')
console.log(highestValuePalindrome('932239', 6, 2))
console.log('===================')
console.log(highestValuePalindrome('11119111', 8, 4))
console.log('===================')

// const result = highestValuePalindrome(test_case_palindrome[0], test_case_palindrome[0].length, 58343)
// let count = 0
// for (let i = 0; i < result.length; i++) {
//     const char = result[i]
//     // if (char !== '9') {
//     //     count++
//     // }
//     const charTwo = test_case_palindrome_result[0][i]
//     if (char !== charTwo) {
//         // console.log(char, charTwo, i)
//         // break
//         count++
//     }
// }
// console.log('count:', count)

// const sliced = test_case_palindrome_result[0].substr(26070, 500)
// const slicedTwo = result.substr(26070, 500)
// const sliced = test_case_palindrome_result[0].substr(11720, 500)
// const slicedTwo = result.substr(11720, 500)
// console.log(sliced)
// console.log(slicedTwo)